pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -235,32 +250,89 @@ func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.Node
 	return potentialNodes
 }
 
+// calculateNumCandidates returns the number of candidates the findCandidates
+// method must produce from dry running based on the constraints given by
+// <maxCandidatesAsPercentageOfNodes> and <minCandidates>. The number of
+// candidates will never be greater than <numNodes>.
+func (pl *DefaultPreemption) calculateNumCandidates(numNodes int32) int32 {
+	n := (numNodes * pl.args.MaxCandidatesAsPercentageOfNodes) / 100
+	if n < pl.args.MinCandidates {
+		n = pl.args.MinCandidates
+	}
+	if n > numNodes {
+		n = numNodes
+	}
+	return n
+}
+
+// calculateOffset returns an <offset> and <numCandidates> that the caller must
+// use to find candidates. A variable offset is used as the starting index when
+// dry running preemption so that the same set of nodes don't get preempted
+// every time. If <offset> + <numCandidates> runs past the number of nodes, the
+// caller is expected to wrap around the node list.
+func (pl *DefaultPreemption) calculateOffset(numNodes int32) (int32, int32) {
+	offset := pl.candidateOffset
+	numCandidates := pl.calculateNumCandidates(numNodes)
+	if pl.candidateOffset >= numNodes {
+		// This happens when the cluster is downscaled and candidateOffset is past
+		// the new, reduced number of nodes. Reset the counter and restart from the
+		// beginning.
+		pl.candidateOffset = 0
+	} else {
+		pl.candidateOffset = (pl.candidateOffset + numCandidates) % numNodes
+	}
+	return offset, numCandidates
+}
+
 // dryRunPreemption simulates Preemption logic on <potentialNodes> in parallel,
-// and returns all possible preemption candidates.
-func dryRunPreemption(ctx context.Context, fh framework.PreemptHandle, state *framework.CycleState,
-	pod *v1.Pod, potentialNodes []*framework.NodeInfo, pdbs []*policy.PodDisruptionBudget) []Candidate {
-	var resultLock sync.Mutex
-	var candidates []Candidate
+// and returns preemption candidates. The number of candidates depends on the
+// constraints defined in the plugin's args. In the returned list of
+// candidates, ones that do not violate PDB are preferred over ones that do.
+func (pl *DefaultPreemption) dryRunPreemption(ctx context.Context, fh framework.PreemptHandle,
+	state *framework.CycleState, pod *v1.Pod, potentialNodes []*framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget) []Candidate {
+	offset, numCandidates := pl.calculateOffset(int32(len(potentialNodes)))
+	var mu sync.Mutex
+	var nonViolatingCandidates []Candidate
```

Comment:
worth initializing with capacity? Do only if `append` shows significant overhead in the profiling.